File: go/glome/glome.go
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package glome implements GLOME protocol.
package glome

import (
	"fmt"
	"io"

	"crypto/hmac"
	"crypto/sha256"

	"golang.org/x/crypto/curve25519"
)

const (
	// PrivateKeySize is the size of a PrivateKey in bytes.
	PrivateKeySize = 32
	// PublicKeySize is the size of a PublicKey in bytes.
	PublicKeySize = 32
	// MaxTagSize is the maximum size allowed for a Tag
	MaxTagSize = 32
	// MinTagSize is the minimum size allowed for a Tag
	MinTagSize = 1
)

var (
	// ErrInvalidPublicKey denotes that a slice that intend to be a public key is not of desired length
	ErrInvalidPublicKey = fmt.Errorf("invalid public key - byte slice len is not %d", PublicKeySize)
	// ErrInvalidPrivateKey denotes that a slice that intend to be a private key is not of desired length
	ErrInvalidPrivateKey = fmt.Errorf("invalid private key - byte slice len is not %d", PrivateKeySize)
	// ErrInvalidTagSize denotes that provided integer is not suitable to be minPeerTagSize
	ErrInvalidTagSize = fmt.Errorf("invalid tag size - minPeerTagSize must be in range %d-%d",
		MinTagSize, MaxTagSize)
	// ErrInvalidReader denotes that library failed to read PrivateKeySize bytes from given Reader.
	ErrInvalidReader = fmt.Errorf("invalid reader - failed to read %d bytes", PrivateKeySize)
)

// PublicKey is the type of GLOME public Keys.
//
// It can be initialized either by casting a [PublicKeySize]byte array or from a byte
// slice with the PublicKeyFromSlice function.
// Examples:
// - Generate Public Key as existing byte array
//       b := [32]byte{0,2,...,7,6}
//       p := glome.PublicKey(b)
//
// - Generate from byte slice
//       s := b[:]
//       p, err := glome.PublicKeyFromSlice(s)
//       if err != nil { [...] }
//
// - Read from File
//       p, err := ioutil.ReadFile(filename)
//       if err != nil { [...] }
//       priv, err := glome.PublicKeyFromSlice(p)
//       if err != nil { [...] }
type PublicKey [PublicKeySize]byte

// PublicKeyFromSlice generates a PublicKey object from slice. Return ErrInvalidPublicKey
// if slice's length is not PublicKeySize.
func PublicKeyFromSlice(b []byte) (*PublicKey, error) {
	if len(b) != PublicKeySize {
		return nil, ErrInvalidPublicKey
	}

	var p PublicKey
	copy(p[:], b)
	return &p, nil
}

// PrivateKey is the type of GLOME public keys.
//
// It can be initialized either by casting a [PrivateKeySize]byte array or from a byte
// slice with the PrivateKeyFromSlice function.
//
// Examples:
// - Generate Private Key as existing byte array:
//       b := [32]byte{0,2,...,7,6}
//       p := glome.PrivateKey(b)
//
// - Generate from byte slice:
//       s := b[:]
//       p, err := glome.PrivateKeyFromSlice(s)
//       if err != nil { [...] }
//
// - Read from File:
//       p, err := ioutil.ReadFile(filename)
//       if err != nil { [...] }
//       priv, err := glome.PrivateKeyFromSlice(p)
//       if err != nil { [...] }
type PrivateKey [PrivateKeySize]byte

// Public returns the PublicKey corresponding to priv.
func (priv *PrivateKey) Public() (*PublicKey, error) {
	slice, err := curve25519.X25519(priv[:], curve25519.Basepoint)
	if err != nil {
		return nil, err
	}

	p, _ := PublicKeyFromSlice(slice)
	return p, nil
}

// Exchange generates a Dialog struct. It performs GLOME handshake, and stores create
// a Dialog from the user to the peer. Sets minPeerTagSize as MaxTagSize.
func (priv *PrivateKey) Exchange(peer *PublicKey) (*Dialog, error) {
	s, err := curve25519.X25519(priv[:], peer[:])
	if err != nil {
		return nil, err
	}

	public, err := priv.Public()
	if err != nil {
		return nil, err
	}

	return &Dialog{shared: s, User: *public, Peer: *peer, minPeerTagSize: MaxTagSize}, nil
}

// TruncatedExchange generates a Dialog struct. It performs GLOME handshake,
// and stores create a Dialog from the user to the peer. Sets param m as minPeerTagSize.
func (priv *PrivateKey) TruncatedExchange(peer *PublicKey, m uint) (*Dialog, error) {
	if m == 0 || m > MaxTagSize {
		return nil, ErrInvalidTagSize
	}

	d, err := priv.Exchange(peer)
	if err != nil {
		return nil, err
	}

	d.minPeerTagSize = m
	return d, nil
}

// PrivateKeyFromSlice generates a private key from a slice. Fail if len of
// slice is not PrivateKeySize
func PrivateKeyFromSlice(b []byte) (*PrivateKey, error) {
	if len(b) != PrivateKeySize {
		return nil, ErrInvalidPrivateKey
	}

	var p PrivateKey
	copy(p[:], b)
	return &p, nil
}

// GenerateKeys generates a public/private key pair using entropy from rand.
func GenerateKeys(rand io.Reader) (*PublicKey, *PrivateKey, error) {
	b := make([]byte, PrivateKeySize)
	n, err := rand.Read(b)
	if err != nil {
		return nil, nil, err
	}
	if n != PrivateKeySize {
		return nil, nil, ErrInvalidReader
	}

	priv, err := PrivateKeyFromSlice(b)
	if err != nil {
		return nil, nil, err
	}

	pub, err := priv.Public()
	if err != nil {
		return nil, nil, err
	}

	return pub, priv, nil
}

// Dialog allow tag managing functionalities for GLOME protocol.
//
// Has to be generated with the methods Exchange or TruncatedExchange or Private key.
// For example:
//       pubKey, privKey, err := glome.GenerateKeys(rand.Reader)
//       if err != nil { [...] }
//       ex, err := privkey.Exchange(peerKey)
//
// If TruncatedExchange is selected, minPeerTagSize can be different to MaxTagSize. See
// documentation in method Check for more information on truncation.
type Dialog struct {
	shared         []byte
	User           PublicKey // User's Public key
	Peer           PublicKey // Peer's Public key
	minPeerTagSize uint      // Minimun Tag size allowed.
}

func (d *Dialog) sendingKey() []byte {
	return append(d.shared[:], append(d.Peer[:], d.User[:]...)...)
}

func (d *Dialog) receivingKey() []byte {
	return append(d.shared[:], append(d.User[:], d.Peer[:]...)...)
}

// Generates a tag matching some provided message, counter and password.
func generateTag(msg []byte, counter uint8, password []byte) []byte {
	h := hmac.New(sha256.New, password)
	h.Write([]byte{counter})
	h.Write(msg)
	return h.Sum(nil)
}

// Tag generates a tag matching some provided message and counter.
// This tag is generated following GLOME protocol specification
// in the context of a communication from the users to theirs peers.
func (d *Dialog) Tag(msg []byte, counter uint8) []byte {
	return generateTag(msg, counter, d.sendingKey())
}

// Check method checks if a tag matches some provided message and counter.
// The method generates the matching tag following GLOME protocol
// specification in the context of a communication from the users'
// peers to the users and then is compared with the tag provided.
//
// For the tag to be accepted it has to be equal in all its length
// to the correct tag. Also, its length must be at least MinPeerTagLength
// and always smaller than MaxTagSize.
func (d *Dialog) Check(tag []byte, msg []byte, counter uint8) bool {
	var prefixSize uint
	switch {
	case uint(len(tag)) < d.minPeerTagSize:
		prefixSize = d.minPeerTagSize
	case uint(len(tag)) > MaxTagSize:
		prefixSize = MaxTagSize
	default:
		prefixSize = uint(len(tag))
	}

	want := generateTag(msg, counter, d.receivingKey())[:prefixSize]
	return hmac.Equal(want, tag)
}

File: python/pyglome/glome.py
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Python GLOME library.

This module contains the Glome class and generate_keys function.

Example use: Sender

>>> import pyglome
>>> tag_manager = pyglome.Glome(peer_key)
>>> first_tag = tag_manager.tag(first_msg, 0) # 0 as it is the first msg
>>> second_tag = tag_manager.tag(second_msg, 1)

Example use: Receiver

>>> import pyglome
>>> tag_manager = pyglome.Glome(peer_key, my_private_key)
>>> ## Need to have a private key (paired to the public key
>>> ## that the sender use)
>>> try:
...     tag_manager.check(tag, msg, counter=0):
>>> except pyglome.IncorrectTagError as wte:
...     ## Handle the exception
>>> ## do what you have to do
"""

import os
import hashlib
import hmac
from typing import NamedTuple

from cryptography.hazmat.primitives.asymmetric import x25519
from cryptography.hazmat.primitives import serialization


class KeyPair(NamedTuple):
    """
    NamedTuple-Class that stores a private/public key pair.

    Attributes:
        - private: A private key.
        - public: A public key paired with the private one.
    """
    private: x25519.X25519PrivateKey
    public: x25519.X25519PublicKey


class Error(Exception):
    """Error super-class for any error that is thrown in PyGLOME."""


class TagCheckError(Error):
    """Raised whenever a tag is not correct or the method failed to check it."""


class IncorrectTagError(Error):
    """Raised whenever the tag provided does not match the message and counter."""


class TagGenerationError(Error):
    """Raised whenever a tag could not be generated."""


def _public_key_encode(public_key: x25519.X25519PublicKey):
    return public_key.public_bytes(serialization.Encoding.Raw,
                                   serialization.PublicFormat.Raw)


def _tag(msg: bytes, counter: int, key: bytes) -> bytes:
    if counter < 0 or counter > 255:
        raise ValueError(
            'Counters for tags must be in range 0-255, not {}'.format(counter))

    message = bytes([counter]) + msg  # msg: N_x|M_n
    digester = hmac.new(key=key, msg=message, digestmod=hashlib.sha256)
    return digester.digest()


class Glome:
    """Implement tag managing functionalities for GLOME protocol.

    This class is initialized by providing your peer's public key and
    optionally your private key. If a private key is not provided, one is
    automatically generated making use of `generate_keys`. Provides methods
    tag (to generate new tags) and check (to check receiving tags).
    """

    MAX_TAG_LEN = 32  # 32 is maximum tag length
    MIN_TAG_LEN = 1

    def __init__(self,
                 peer_key: x25519.X25519PublicKey,
                 my_private_key: x25519.X25519PrivateKey = None,
                 min_peer_tag_len: int = MAX_TAG_LEN):
        """Initialize Glome class.

        Performs the handshake and generates keys.

        Args:
            peer_key: Your peer's public key.
            my_private_key: Your private key.
            min_peer_tag_len: Desired length (in bytes) for the tag.
              Must be an integer in range 1-32.
        Raises:
            ValueError: Raised whenever min_peer_tag_len is not in
              range 1-32.
        """

        if my_private_key is None:
            my_private_key, my_public_key = generate_keys()
        else:
            my_public_key = my_private_key.public_key()

        if not Glome.MIN_TAG_LEN < min_peer_tag_len <= Glome.MAX_TAG_LEN:
            raise ValueError('min_peer_tag_len must be in range {}-{}'.format(
                Glome.MIN_TAG_LEN, Glome.MAX_TAG_LEN))

        shared_secret = my_private_key.exchange(peer_key)

        self._send_key = shared_secret + _public_key_encode(
            peer_key) + _public_key_encode(my_public_key)
        self._receive_key = shared_secret + _public_key_encode(
            my_public_key) + _public_key_encode(peer_key)
        self._peer_key = peer_key
        self._my_keys = KeyPair(my_private_key, my_public_key)
        self._min_peer_tag_len = min_peer_tag_len

    @property
    def user_keys(self) -> KeyPair:
        """User's private and public keys used in handshake."""
        return self._my_keys

    @property
    def peer_key(self) -> x25519.X25519PublicKey:
        """Peer's public key used in handshake."""
        return self._peer_key

    def tag(self, msg: bytes, counter: int) -> bytes:
        """Generates a tag from a message and a counter.

        Generates a tag matching some provided message and counter.
        This tag is generated following GLOME protocol specification
        in the context of a communication from the users to theirs peers.

        Args:
           msg: Message to be transmitted.
           counter: Numbers of messages transmitted previously in the
             conversation in this direction (i.e. from the user
             to the peer). Must be an integer in {0,...,255}.
        Returns:
           tag: Tag matching counter and msg.
        Raises:
           TagGenerationError: Raised whenever the method failed to generate tag
             due to ValueError in the arguments.
        """
        try:
            return _tag(msg, counter, self._send_key)
        except ValueError as value_error:
            raise TagGenerationError('Failed to generate tag') from value_error

    def check(self, tag: bytes, msg: bytes, counter: int):
        """Check whether a tag is correct for some message and counter.

        Checks if a tag matches some provided message and counter.
        The method generates the matching tag following GLOME protocol
        specification in the context of a communication from the users'
        peers to the users and then is compared with the tag provided.

        Args:
           tag: Object with the generated tag.
           msg: Object containing received message.
           counter: Numbers of messages transmitted previously in the
             conversation in this direction (i.e. from the peer
             to the user).
        Returns:
           None.
        Raises:
           TagCheckError: Raised whenever the method fails to check the tag
             due to a ValueError in the arguments.
           IncorrectTagError: Raised whenever the tag is incorrect.
        """
        prefix_length = max(len(tag), self._min_peer_tag_len)
        prefix_length = min(prefix_length, Glome.MAX_TAG_LEN)

        try:
            correct_tag = _tag(msg, counter, self._receive_key)[:prefix_length]
        except ValueError as value_error:
            raise TagCheckError('Failed to check the tag') from value_error

        if not hmac.compare_digest(tag, correct_tag):
            raise IncorrectTagError('Tag provided does not match correct tag')


def generate_keys() -> KeyPair:
    """Generates a private/public key pair.

    Provides a random key pair based output of os.urandom. The format
    matches the one requested by Glome Class.

    Args:
       None
    Returns:
       A KeyPair, containing a random private key and the public key derived
       from the generated private key
    """
    private = x25519.X25519PrivateKey.from_private_bytes(
        os.urandom(Glome.MAX_TAG_LEN))
    return KeyPair(private, private.public_key())


class AutoGlome:
    """Adds counter managing functionalities for GLOME protocol.

    This class is initialized by providing your peer's public key and
    optionally your private key. If a private key is not provided, one is
    automatically generated making use of `generate_keys`. On initialization,
    two counter (sending and receiving) are created and set to 0. Provides
    methods tag (to generate new tags) and check (to check receiving tags).
    """

    def __init__(self,
                 peer_key: x25519.X25519PublicKey,
                 my_private_key: x25519.X25519PrivateKey = None,
                 *,
                 min_peer_tag_len: int = Glome.MAX_TAG_LEN,
                 skippable_range: int = 0):
        """Initialize AutoGlome class.

        Performs the handshake, generates keys and counters.
        Args:
           peer_key: Your peer's public key.
           my_private_key: Your private key.
           min_peer_tag_len: Desired length (in bytes) for the tag.
             Must be an integer in range 1-32. keyword only.
           skippable_range: Number of messages that can be missed. keyword only.
             Must be non-negative. For more information please go to check method's
             documentation.
        Raises:
           ValueError: Raised whenever min_peer_tag_len is not in range 1-32 or
             skippable_length is a negative integer.
        """
        if skippable_range < 0:
            raise ValueError(
                'Skippable_range must be non-negative, not {}'.format(
                    skippable_range))

        self.glome = Glome(peer_key,
                           my_private_key,
                           min_peer_tag_len=min_peer_tag_len)
        self._sending_counter = 0
        self._receiving_counter = 0
        self.skippable_range = skippable_range

    @property
    def sending_counter(self) -> int:
        """Number of tags shared from the user to the peer.

        It is incremented each time a new tag is generated. It is always
        one byte long. When the counter gets past 255 it overflows at 0.

        Setter raises ValueError if provided integer is not in range 0-255.
        """
        return self._sending_counter

    @sending_counter.setter
    def sending_counter(self, value: int):
        if not 0 <= value <= 255:
            raise ValueError('Counter must be in range 0-255')
        self._sending_counter = value

    @property
    def receiving_counter(self) -> int:
        """Number of tags the user receives from the peer.

        It is always one byte long. When the counter gets past 255 it restarts at
        0. Every time a message is successfully checked, the receiving_counter is
        set to the next value after the last successful one. Note that if
        skippable_range is n the counter might be increased by any amount in
        range 1-n+1 after a successful check.

        Setter raises ValueError if provided counter is not in range 0-255.
        """
        return self._receiving_counter

    @receiving_counter.setter
    def receiving_counter(self, value: int):
        if not 0 <= value <= 255:
            raise ValueError('Counter must be in range 0-255')
        self._receiving_counter = value

    @property
    def user_keys(self) -> KeyPair:
        """User's private and public keys used in handshake."""
        return self.glome.user_keys

    @property
    def peer_key(self) -> x25519.X25519PublicKey:
        """Peer's public key used in handshake."""
        return self.glome.peer_key

    def tag(self, msg: bytes) -> bytes:
        """Generates a tag from a message.

        Generates a tag matching some provided message and the internal
        sending counter. This tag is generated following GLOME protocol
        specification in the context of a communication from the users to
        theirs peers.

        Args:
           msg: Message to be transmitted.
        Returns:
           tag: Tag matching counter and msg.
        Raises:
           TagGenerationError: Raised whenever the method failed to generate tag
             due to ValueError in the arguments.
        """
        tag = self.glome.tag(msg, self.sending_counter)
        self._sending_counter = (self._sending_counter + 1) % 256
        return tag

    def check(self, tag: bytes, msg: bytes):
        """Check whether a tag is correct for some message.

        Checks if a tag matches some provided message and internal receiving
        counter. The method generates the matching tag following GLOME protocol
        specification in the context of a communication from the users' peers to
        the users and then is compared with the tag provided. If tag checking if
        not successful, the receiving counter remains unchanged.

        If skippable_range if greater than 0, the method try to check the tag
        against all counters in range [receiving_counter, receiving_counter +
        skippable_range], in order, until one is successful. If no one is successful,
        an exceptions is raised and receiving counter remains unchanged.

        Args:
           tag: Object with the generated tag.
           msg: Object containing received message.
        Returns:
           None.
        Raises:
           IncorrectTagError: Raised whenever the tag is incorrect.
        """
        old_counter = self._receiving_counter
        for _ in range(self.skippable_range + 1):
            try:
                self.glome.check(tag, msg, self.receiving_counter)
                self._receiving_counter = (self._receiving_counter + 1) % 256
                return None
            except IncorrectTagError:
                self._receiving_counter = (self._receiving_counter + 1) % 256

        #If no counter matches.
        self._receiving_counter = old_counter
        raise IncorrectTagError('Tag provided does not match correct tag')
